enhancements/security/scc-run-as-any.md

```diff
@@ -0,0 +1,134 @@
+---
+title: scc-run-as-any
+authors:
+  - "@alanconway"
+reviewers:
+  - "@eparis"
+  - "@bparees"
+  - "@jcantrill"
+approvers:
+  - "@eparis"
+api-approvers:
+  - None
+creation-date: 2023-05-17
+last-updated:
+tracking-link:
+  - TBD
+---
+
+# Consistent handling for SecurityContextConstraint RunAsAny strategies
+
+## Summary
+
+SecurityContextConstraints serve two distinct functions:
+
+1. They _constrain_ Pod configurations that are allowed to run.
+2. They _modify_ Pod configurations with values from outside the Pod based on strategies in the SCC.
+
+From [the docs][scc]:
+> The admission controller is aware of certain conditions in the security context constraints (SCCs) that trigger it to look up pre-allocated values from a namespace and populate the SC before processing the pod.
+
+SCC strategies are ordered by "restrictiveness": if a Pod runs under SCC X, then it MUST run under any SCC that is _less restrictive_ than X.
+This ordering is critical to correctly matching SCCs to Pods, and must be clear an consistent.
+
+The `RunAsAny` strategy is documented as being the _least restrictive_ `runAsUser` strategy.
+In other words, a Pod that runs under some other `runAsUser` strategy must also run under `RunAsAny`
+
+This is not currently the case for pods with `runAsNonRoot:true`. They run correctly under `MustRunAsRange` but fail under `RunAsAny`.
+If a Pod has no `runAsUser` and no image `User` setting then:
+- Under `MustRunAsRange`, SCC processing "helps" by providing a default non-root UID from a predefined range.
+- Under `RunAsAny`, SCC processing does not help, and assumes the Docker default of 0. The Pod is rejected by its own security context.
+
+The proposal: For `runAsNonRoot:true` pods, `RunAsAny` should use the same defaulting rules as `MustRunAsRange` to provide a non-0 UID.
```

Comment:
> for existing workloads that:
> 
>     set runAsNonRoot:true
>     run an image that has a non-zero default uid
> 
> this would change what uid the container runs under, which in turn might break those containers if they need to run as a specific uid.
> 

I'll clarify the proposal I intended to state that there is NO CHANGE unless all of the following hold:
- runAsNonRoot:true
- There no runAs* field
- The image has User=""  (an explicit User="0" should cause failure as it's an explicit mis-configuration) 

The proposal is intended to apply _only_ when the Pod gives absolutely no indication of preference for a UID _except_ runAsNonRoot.

Reply:
it's still a change in behavior.  you're going to go from not running the container, to running it with some new uid.

also you can't know if the image has a non-root default user at the layer that needs to make this decision, unless you're proposing that the container runtime itself on the node makes the decision to run the container with a range uid, which then means it needs to know what the valid range is, so it's moving responsibilities around quite a bit.

Reply:
So how does MustRunAsRange work? It claims to check that the container will run as a UID in the range, and it adds a runAsUser if there is no default. I'm simply proposing to make RunAsAny use the same defaulting rules that MustRunAsRange already implements. I'm working on a clearer wording ass the pieces are falling into place in my head.

Reply:
> So how does MustRunAsRange work? It claims to check that the container will run as a UID in the range, and it adds a runAsUser if there is no default. 

i believe that's correct.  And will also reject the pod if the pod specifies a UID that is not in the range.

>  I'm simply proposing to make RunAsAny use the same defaulting rules that MustRunAsRange already implements

but that is a change, because today RunAsAny does not set a runasuser on the pod.  If the pod doesn't specify one, runasany still admits the pod (because if you can use that SCC, you're allowed to run pods as any user, so it doesn't matter whether you specify a user on the pod or not, and it doesn't matter what user the image run as.....from the perspective of the RunAsAny permission. (the RunAsNonRoot permission is separate and can still block you at runtime, as you know now).

So again, you'd be changing the behavior.  Today a pod that:
1) sets runasNonroot=true
2) doesn't set a runasuser
3) gets admitted under a RunAsAny SCC

will run as the default image uid (which may or may not fail depending if that default uid is root)

w/ your proposal as i understand it, the same pod would get assigned a uid in the range.  Meaning the image will *not* attempt to run as the default image uid.  Which means either you've changed the behavior of the pod/image from "fails to run" to "run as range uid", or you've changed it from "runs as default image uid" to "runs as range uid".

both of those outcomes are changes in behavior.

Reply:
>  Which means either you've changed the behavior of the pod/image from "fails to run" to "run as range uid", or you've changed it from "runs as default image uid" to "runs as range uid".

Correct. "Fails to run" is the problem I am trying to address. The new semantics for "run as range UID" are identical to the existing semantics of MustRunAsRange. I'm proposing to make the behaviour consistent, and to fix the behavior where a functioning system can be crashed by introducing a new SCC that with higher priority that is  _less restritive_ than the SCC that previously matched.  To me, the current definition of _less restrictive_ makes it impossible to reason about the effects of adding a new SCC, as demonstrated by the customer outage.

Reply:
Changing the behavior of a workload between cluster versions given the same inputs/config/api field values is fundamentally not something we tend to allow without really good reason(because it's effectively an api change). I don't think this particular situation warrants it. Hence why I'm pushing for other solutions to this problem which would allow the user to better specify their desired intent, rather than changing the behavior for existing workload configurations, whether or not we think most people might prefer the new behavior.